### PR TITLE
Allows one to push the firehose-to-syslog nozzle as an app.

### PR DESCRIPTION
**Pull request by malston**
Useful in environments not using the logsearch-for-cloudfoundry bosh release.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -12,20 +12,22 @@ import (
...
+	subscriptionId    = kingpin.Flag("subscription-id", "Id for the subscription.").Default("firehose").OverrideDefaultFromEnvar("FIREHOSE_SUBSCRIPTION_ID").String()
+	user              = kingpin.Flag("user", "Admin user.").Default("admin").OverrideDefaultFromEnvar("FIREHOSE_USER").String()
+	password          = kingpin.Flag("password", "Admin password.").Default("admin").OverrideDefaultFromEnvar("FIREHOSE_PASSWORD").String()
+	skipSSLValidation = kingpin.Flag("skip-ssl-validation", "Please don't").Default("false").OverrideDefaultFromEnvar("SKIP_SSL_VALIDATION").Bool()
+	cfPush            = kingpin.Flag("cf-push", "Deploy to Cloud Foundry.").Default("false").OverrideDefaultFromEnvar("CF_PUSH").Bool()
+	wantedEvents      = kingpin.Flag("events", fmt.Sprintf("Comma separated list of events you would like. Valid options are %s", events.GetListAuthorizedEventEvents())).Default("LogMessage").OverrideDefaultFromEnvar("EVENTS").String()
 	boltDatabasePath  = kingpin.Flag("boltdb-path", "Bolt Database path ").Default("my.db").String()
 	tickerTime        = kingpin.Flag("cc-pull-time", "CloudController Pooling time in sec").Default("60s").Duration()
```
Comment (shinji62) [inline main.go]: Can be nice to be able to configure with env var.

File: main.go [hunk 2]
```diff
@@ -129,3 +135,24 @@ func main() {
...
+	fmt.Fprintln(w,
+		"Hello!\nWe have processed", events.GetTotalCountOfSelectedEvents(), "total events.",
+		"\nWe have tapped the firehose at", fmt.Sprintf("wss://doppler.%s", *dopplerEndpoint))
+	fmt.Fprintln(w, "\nOf those events, we have processed\n")
+	for event, count := range events.GetSelectedEventsCount() {
+		fmt.Fprintln(w, count, event, "events\n")
+	}
+}
+
+func setupHTTP() {
```
Comment (shinji62) [inline main.go]: Do we really need the http server ? What about injecting this information into the output string every X time.
We will be able to visualize into ELK.

**Comment by simonjohansson** [comment]
Hi, thanks for the pr @malston !

I will look over this, but as an initial impression Im with @shinji62 , we should be able to run this in CF with --no-route? I will look into that.

File: main.go (hunk 1 above)
Comment (simonjohansson) [inline main.go]: Agreed, I think we should allow all config options to be overridden by env variables.

* boltDatabasePath
* tickerTime
* extraFields
* modeProf
* pathProf

**Comment by simonjohansson** [comment]
```
applications:
- name: firehose-to-syslog
  buildpack: https://github.com/cloudfoundry/go-buildpack.git
  env: 
    API_ENDPOINT: https://api.domain.com
    FIREHOSE_USER: admin
    FIREHOSE_PASSWORD: kehe-super-secret
    DEBUG: true
```

cf push --no-route works like a charm without any need for a http server.

I think its better to output the data every n seconds like

```
firehoseStatus := time.NewTicker(*ticker)

go func() {
    for range firehoseStatus.C {
        logging.LogStd(getInfo())
    }
}()
```

Any particular reason why you chose to expose this over HTTP @malston ?

**Comment by malston** [comment]
The only reason was to show some output to the user in case they wanted to check the status. But that's not as important as being able to scale it. So I'm happy to remove the http endpoint.

**Comment by shinji62** [comment]
Nice,
Before merging 
* Can you make the changes asked by Simon for the configurable var ? 
* Can you give use an sample of the output of event count (use --debug)?


Somethings comes into my mind : 
* Total events is fine from the starting of the nozzle ==> Already implemented 
* Event by sec, deducting the event by sec based on the ticker time and so on. ==> What about this one  ?

**Comment by malston** [comment]
Thanks @shinji62 
I'm still working on the output formatting of the events totals. Should we put this all on one line? Is it worthwhile breaking it down by event type?

**Comment by simonjohansson** [comment]
@malston 

I think one line per event type, makes it easier to filter over time in Logstash.

**Comment by malston** [comment]
Here's some sample output from tailing the logs of the app.  I appended (firehose-to-syslog-events) to the message to make it easier to grep.
```
○ → cf logs firehose-to-syslog | egrep -h '(firehose-to-syslog-events)'
2015-08-28T12:58:18.60-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":"(firehose-to-syslog-events) We have processed 653973 total events from the firehose at wss://doppler.run.pez.pivotal.io:443","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-08-28T18:58:18Z","timestamp":1440788254372396165}
2015-08-28T12:59:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 985880 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T12:59:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) And of those events, we have processed
2015-08-28T12:59:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) 897811 LogMessage
2015-08-28T12:59:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) 81436 ValueMetric
2015-08-28T12:59:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) 68 ContainerMetric
2015-08-28T12:59:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1000156 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T12:59:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) And of those events, we have processed
2015-08-28T12:59:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 6752 HttpStartStop
2015-08-28T12:59:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 36 CounterEvent
2015-08-28T12:59:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 70 ContainerMetric
2015-08-28T13:00:34.38-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1103083 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:00:34.38-0600 [App/0]      OUT (firehose-to-syslog-events) And of those events, we have processed
2015-08-28T13:00:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1117423 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:00:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) And of those events, we have processed
2015-08-28T13:00:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) 1015845 LogMessage
2015-08-28T13:00:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) 93884 ValueMetric
2015-08-28T13:00:44.37-0600 [App/0]      OUT (firehose-to-syslog-events) 38 CounterEvent
2015-08-28T13:00:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1137788 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:00:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 1034849 LogMessage
2015-08-28T13:00:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 7654 HttpStartStop
2015-08-28T13:00:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 39 CounterEvent
2015-08-28T13:01:04.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1158184 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:01:04.37-0600 [App/0]      OUT (firehose-to-syslog-events) And of those events, we have processed
2015-08-28T13:01:04.37-0600 [App/0]      OUT (firehose-to-syslog-events) 97801 ValueMetric
2015-08-28T13:01:10.78-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":"(firehose-to-syslog-events) 82 ContainerMetric","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-08-28T19:01:10Z","timestamp":1440788464372352189}
2015-08-28T13:01:14.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1177635 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:01:14.37-0600 [App/0]      OUT (firehose-to-syslog-events) 7987 HttpStartStop
2015-08-28T13:01:14.37-0600 [App/0]      OUT (firehose-to-syslog-events) 41 CounterEvent
2015-08-28T13:01:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1284360 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:01:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) And of those events, we have processed
2015-08-28T13:01:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 107568 ValueMetric
2015-08-28T13:01:54.37-0600 [App/0]      OUT (firehose-to-syslog-events) 91 ContainerMetric
2015-08-28T13:02:03.87-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":"(firehose-to-syslog-events) And of those events, we have processed","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-08-28T19:02:03Z","timestamp":1440788514372344474}
2015-08-28T13:02:03.87-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":"(firehose-to-syslog-events) 107568 ValueMetric","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-08-28T19:02:03Z","timestamp":1440788514372386469}
2015-08-28T13:02:14.37-0600 [App/0]      OUT (firehose-to-syslog-events) We have processed 1309432 total events from the firehose at wss://doppler.run.pez.pivotal.io:443
2015-08-28T13:02:14.37-0600 [App/0]      OUT (firehose-to-syslog-events) 46 CounterEvent
2015-08-28T13:02:14.37-0600 [App/0]      OUT (firehose-to-syslog-events) 92 ContainerMetric
```

**Comment by simonjohansson** [comment]
Looks neato.

Just a couple of last notes

* "(firehose-to-syslog-events)", whilst a good idea for greping in bash would be a bit annoying when looking at the production logs, when not run in debug mode only these messages would be printed to stdout not the events themselves.

* could we make dur configurable like https://github.com/cloudfoundry-community/firehose-to-syslog/blob/develop/main.go#L30 ?

* could we log the number of processed events since the last tick? That way it would be easier to spot anomalies, and draw fancy graphs.
```
2015-08-28T13:01:14.37-0600 [App/0]      OUT We have processed xxx total events from the firehose at wss://doppler.run.pez.pivotal.io:443 the last N seconds and yyy since startup
...
2015-08-28T13:01:54.37-0600 [App/0]      OUT yyy ValueMetric the last N seconds
``` 

Besides that everything looks groovy to me :+1:

**Comment by malston** [comment]
Thanks @simonjohansson for the quick feedback -- I'll get these changes in shortly.

**Comment by simonjohansson** [comment]
@malston great! Ill have a look at this today. :)

**Comment by shinji62** [comment]
will check too.

**Comment by simonjohansson** [comment]
@malston could you move getEventTotals and logEventTotals into the logging package, that way we keep as much implementation details as possible out of main :)

Also, the logging creates a big event with uneccecary newlines 

```
[2015-09-03 10:56:30.41167407 +0200 CEST]  
We have processed 288 events from the firehose at  over the last 30 seconds and 288 total events since startup
 
And of those events, we have processed

 
 288 CounterEvent over the last 30 seconds


```

Could we make this into 

```
[2015-09-03 10:56:30.41167407 +0200 CEST] We have processed 288 events from the firehose at  over the last 30 seconds and 288 total events since startup
[2015-09-03 10:56:30.41167407 +0200 CEST] And of those events, we have processed
[2015-09-03 10:56:30.41167407 +0200 CEST] 288 CounterEvent over the last 30 seconds 
```

**Comment by malston** [comment]
I get the following output when I tail the logs:
```
±  |malston-feature/cf-push ✗| → cf logs firehose-to-syslog | egrep -h 'seconds'
2015-09-03T10:25:03.69-0600 [App/0]      OUT We have processed 6916 events from the firehose at wss://doppler.run.pez.pivotal.io:443 over the last 10 seconds and 6916 total events since startup
2015-09-03T10:25:03.69-0600 [App/0]      OUT  4599 LogMessage over the last 10 seconds
2015-09-03T10:25:03.69-0600 [App/0]      OUT  331 HttpStartStop over the last 10 seconds
2015-09-03T10:25:03.69-0600 [App/0]      OUT  1985 ValueMetric over the last 10 seconds
2015-09-03T10:25:03.69-0600 [App/0]      OUT  1 ContainerMetric over the last 10 seconds
2015-09-03T10:25:13.68-0600 [App/0]      OUT We have processed 3325 events from the firehose at wss://doppler.run.pez.pivotal.io:443 over the last 9.9 seconds and 10241 total events since startup
2015-09-03T10:25:13.68-0600 [App/0]      OUT  6122 LogMessage over the last 20 seconds
2015-09-03T10:25:13.68-0600 [App/0]      OUT  562 HttpStartStop over the last 20 seconds
2015-09-03T10:25:13.68-0600 [App/0]      OUT  3555 ValueMetric over the last 20 seconds
2015-09-03T10:25:13.68-0600 [App/0]      OUT  1 ContainerMetric over the last 20 seconds
2015-09-03T10:25:13.68-0600 [App/0]      OUT  1 CounterEvent over the last 20 seconds
2015-09-03T10:25:19.67-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":" 4599 LogMessage over the last 10 seconds","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:19Z","timestamp":1441297503694383493}
2015-09-03T10:25:19.67-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":" 331 HttpStartStop over the last 10 seconds","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:19Z","timestamp":1441297503694395965}
2015-09-03T10:25:23.68-0600 [App/0]      OUT We have processed 11303 events from the firehose at wss://doppler.run.pez.pivotal.io:443 over the last 9.9 seconds and 21544 total events since startup
2015-09-03T10:25:23.68-0600 [App/0]      OUT  14255 LogMessage over the last 30 seconds
2015-09-03T10:25:23.68-0600 [App/0]      OUT  1156 HttpStartStop over the last 30 seconds
2015-09-03T10:25:23.68-0600 [App/0]      OUT  4 CounterEvent over the last 30 seconds
2015-09-03T10:25:19.76-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":"We have processed 6916 events from the firehose at wss://doppler.run.pez.pivotal.io:443 over the last 10 seconds and 6916 total events since startup","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:19Z","timestamp":1441297503694323568}
2015-09-03T10:25:19.76-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":" 1985 ValueMetric over the last 10 seconds","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:19Z","timestamp":1441297503694407368}
2015-09-03T10:25:33.68-0600 [App/0]      OUT  16383 LogMessage over the last 40 seconds
2015-09-03T10:25:33.68-0600 [App/0]      OUT  1808 HttpStartStop over the last 40 seconds
2015-09-03T10:25:33.68-0600 [App/0]      OUT  9789 ValueMetric over the last 40 seconds
2015-09-03T10:25:33.68-0600 [App/0]      OUT  8 CounterEvent over the last 40 seconds
2015-09-03T10:25:35.69-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":" 3555 ValueMetric over the last 20 seconds","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:35Z","timestamp":1441297513689324605}
2015-09-03T10:25:35.69-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":" 1 CounterEvent over the last 20 seconds","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:35Z","timestamp":1441297513689347690}
2015-09-03T10:25:41.44-0600 [App/0]      OUT {"cf_app_id":"c0f14076-66ab-4e17-ba4d-076588a3e35f","cf_app_name":"firehose-to-syslog","cf_org_id":"45f92610-f258-4f6f-9846-67b1cd98aed3","cf_org_name":"sys-pez","cf_origin":"firehose","cf_space_id":"ea88ed9e-91f1-4763-8eef-54fe38acf603","cf_space_name":"pez-dev","event_type":"LogMessage","level":"info","message_type":"OUT","msg":"{\"cf_app_id\":\"c0f14076-66ab-4e17-ba4d-076588a3e35f\",\"cf_app_name\":\"firehose-to-syslog\",\"cf_org_id\":\"45f92610-f258-4f6f-9846-67b1cd98aed3\",\"cf_org_name\":\"sys-pez\",\"cf_origin\":\"firehose\",\"cf_space_id\":\"ea88ed9e-91f1-4763-8eef-54fe38acf603\",\"cf_space_name\":\"pez-dev\",\"event_type\":\"LogMessage\",\"level\":\"info\",\"message_type\":\"OUT\",\"msg\":\" 1985 ValueMetric over the last 10 seconds\",\"origin\":\"dea_logging_agent\",\"source_instance\":\"0\",\"source_type\":\"App\",\"time\":\"2015-09-03T16:25:19Z\",\"timestamp\":1441297503694407368}","origin":"dea_logging_agent","source_instance":"0","source_type":"App","time":"2015-09-03T16:25:41Z","timestamp":1441297519764466832}
2015-09-03T10:25:43.68-0600 [App/0]      OUT We have processed 6405 events from the firehose at wss://doppler.run.pez.pivotal.io:443 over the last 10 seconds and 34405 total events since startup
2015-09-03T10:25:43.68-0600 [App/0]      OUT  8 CounterEvent over the last 50 seconds
2
```
The individual event totals are cumulative. Let me know if you want them to be non-cumulative instead. That'll take some refactoring.

**Comment by simonjohansson** [comment]
Thanks @malston !
Ill merge this first thing in the morning.

Have a nice evening :)

**Comment by shinji62** [comment]
Nice.

**Pull request merged**